### PR TITLE
Add Data Source For Retrieving Schema Names From A Database

### LINKED ISSUE #142: Allow permissions to be applied to all schemas as an option instead of having to pass in a list to account for databases with rapid changing schemas
Hi there,

Thank you for opening an issue. Please provide the following information:

### Terraform Version
14.

I would like to have the https://registry.terraform.io/providers/cyrilgdn/postgresql/latest/docs/resources/postgresql_grant resource the ability to make changes to all schemas in the rds postgresql instance rather than me passing in a list of it. This list of schemas may grow overtime and would like to just re-run terraform instead of updating my list and then re-running or it failing because a schemas was purged. 
The schemas are currently not being managed by TF

### PR DESCRIPTION
**Pull request by SpencerBinXia**
## Overview

This PR adds the new data source `postgresql_schemas` for retrieving schema names from a database. 

The intended usage of this data source is so that it will no longer be necessary to type out schema names explicitly when using resources such as `postgresql_grant` to iterate over a list of schemas within a database. Instead, one may just loop over the list output from `data.postgresql_schemas`.

## Example

```terraform
data "postgresql_schemas" "example_retrieve_schemas" {
    database = "example_database"
    include_system_schemas = true # Optional argument that determines whether to include reserved system schemas.
    like_any_patterns = ["test%","%pg%"] # Optional expressions which will be pattern matched in the query using the LIKE ANY operators. 
    like_all_patterns = ["test%","%pg%"] # Optional expressions which will be pattern matched in the query using the LIKE ALL operators. 
    not_like_all_patterns =  ["example%"] # Optional expressions which will be pattern matched in the query using the NOT LIKE ALL operators. 
    regex_pattern = "^test_schema.*$" # Optional expression which will be pattern matched in the query using the ~ (regular expression match) operator.
}
```

```terraform
output "example_database_schemas" {
    value = data.postgresql_schemas.example_retrieve_schemas.schemas
}
```
## Issues

Closes #142 

This issue was created by another member of our organization.

### CODE REVIEW

**Comment by SpencerBinXia** [comment]
Tests failed the first time due to I believe a different number of temporary pg_ schemas (pg_temp_1, pg_toast_temp_1, etc) than expected from local runs. Changed the acceptance tests to not rely on those, so hopefully second time's the charm?

**Comment by briantist** [comment]
This looks really useful.. any movement on this?

**Comment by SpencerBinXia** [comment]
HI @cyrilgdn , could you please take another look at this MR or let us know any concerns you currently have preventing this from being merged?

**Comment by cyrilgdn** [comment]
@SpencerBinXia I'm having a look, sorry for the delay.

I allowed myself to merge master in your PR.

**Review by cyrilgdn — COMMENTED** [review]
Just a little comment about the `concatenateQueryWithPatternMatching`, let me know what do you thing about that.

Otherwise it seems good to me :+1: 
Sorry again for the delay.

File: postgresql/data_source_postgresql_schemas.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	if len(likeAllPatterns) > 0 {
+		query = concatenateQueryWithPatternMatching(query, likePatternQuery, generatePatternArrayString(likeAllPatterns, queryArrayKeywordAll), &queryContainsWhere)
+	}
+	if len(notLikeAllPatterns) > 0 {
+		query = concatenateQueryWithPatternMatching(query, notLikePatternQuery, generatePatternArrayString(notLikeAllPatterns, queryArrayKeywordAll), &queryContainsWhere)
+	}
+	if regexPattern != "" {
+		query = concatenateQueryWithPatternMatching(query, regexPatternQuery, fmt.Sprintf("'%s'", regexPattern), &queryContainsWhere)
+	}
```
[the remark this answers is not in the document]
Comment (luis-guimaraes-exoawk) [reply postgresql/data_source_postgresql_schemas.go:143]: @SpencerBinXia could you take a look? We would benefit largely if your code got merged!

**Comment by SpencerBinXia** [comment]
Hi @cyrilgdn, I liked your suggestions and made the appropriate changes to remove the need for pointers.

I opened a new PR https://github.com/cyrilgdn/terraform-provider-postgresql/pull/214 that also has changes for adding sequence and table data sources. It would be preferable if possible to get that PR merged and skip this one entirely, but I understand if you want to start with the smaller PR first!

**Comment by tspearconquest** [comment]
Hello any progress on this or #214 please?

**Pull request closed without merging**